lib/Drush/Boot/BootstrapManager.php

```diff
@@ -0,0 +1,147 @@
+<?php
+
+namespace Drush\Boot;
+
+use Psr\Log\LoggerInterface;
+
+class BootstrapManager {
+
+  /**
+   * @var Drush\Boot\Boot[]
+   */
+  protected $bootstrapCandidates = [];
+
+  /**
+   * @var Drush\Boot\Boot
+   */
+  protected $defaultBootstrapObject;
+
+  /**
+   * @var Drush\Boot\Boot
+   */
+  protected $bootstrap;
+
+  /**
+   * @var Psr\Log\LoggerInterface
+   */
+  protected $logger;
+
+  /**
+   * @var string
+   */
+  protected $root;
+
+  /**
+   * @var string
+   */
+  protected $uri;
+
+  /**
+   * Constructor.
+   *
+   * @param Boot
+   *   The default bootstrap object to use when there are
+   *   no viable candidates to use (e.g. no selected site)
+   * @param LoggerInterface
+   *   The logger
+   */
+  public function __construct(Boot $default, LoggerInterface $logger) {
+    $this->defaultBootstrapObject = $default;
+    $this->logger = $logger;
+  }
+
+  /**
+   * Add a bootstrap object to the list of candidates
+   *
+   * @param Boot|Array
+   *   List of boot candidates
+   */
+  public function add($candidateList) {
+    foreach (func_get_args() as $candidate) {
+      $this->bootstrapCandidates[] = $candidate;
+    }
+  }
+
+  /**
+   * Return the framework root selected by the user.
+   */
+  public function getRoot() {
+    return $this->root;
+  }
+
+  public function setRoot($root) {
+    // TODO: Throw if we already bootstrapped a framework?
+    $this->root = $root;
+  }
+
+  /**
+   * Return the framework root selected by the user.
+   */
+  public function getUri() {
+    return $this->uri;
+  }
+
+  public function setUri($uri) {
+    // TODO: Throw if we already bootstrapped a framework?
+    $this->uri = $root;
+  }
+
+  /**
+   * Return the bootstrap object in use.  This will
+   * be the latched bootstrap object if we have started
+   * bootstrapping; otherwise, it will be whichever bootstrap
+   * object is best for the selected root.
+   */
+  public function getBootstrap() {
+    if ($this->bootstrap) {
+      return $this->bootstrap;
+    }
+    return $this->select_bootstrap_class();
```

Comment:
I agree that switching to camel case makes sense. Can be a follow up if desired.